compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5359,6 +5359,15 @@ public void visit(BLangBinaryExpr binaryExpr, AnalyzerData data) {
                 if (leftConstituent != null && rightConstituent != null) {
                     actualType = new BXMLType(BUnionType.create(null, leftConstituent, rightConstituent), null);
                     break;
+                } else if (leftConstituent != null || rightConstituent != null) {
```

Comment:
In this case, you can update the `getXMLConstituents` method to check whether the type is assignable to a string or not. Also, change the function name accordingly to reflect the new logic. By doing this, we can eliminate this logic.

Reply:
I dont think its possible to do this because both types can be string in that case and this only handles xml types